### PR TITLE
feat: integrate builder

### PR DESCRIPTION
**Pull request by mattsse**
Integrates the new builder API.

Removes all extension trait stuff

the launch code is now part of the builder

I've also modified the `receipt_root` functions because mutually exclusive function signatures are a nightmare if we use conflicting ethereum code by default

one example is not completely ported, but I think this didn't even work previously, would like to tackle this in a followup because requires more work.

### CODE REVIEW

File: bin/reth/src/commands/node/mod.rs [hunk 1]
```diff
@@ -128,6 +132,77 @@ pub struct NodeCommand<Ext: RethCliExt = ()> {
...
+        tracing::info!(target: "reth::cli", path = ?db_path, "Opening database");
+        let database = Arc::new(
+            init_db(db_path.clone(), DatabaseArguments::default().log_level(db.log_level.clone()))?
+                .with_metrics(),
+        );
+
+        // set up node config
+        let mut node_config = NodeConfig {
+            // TODO
+            database: DatabaseBuilder::Test,
```
Comment (Rjected) [inline bin/reth/src/commands/node/mod.rs:177]: guessing the TODO is for setting this to `DatabaseBuilder::Real(db)`

Comment (mattsse) [reply bin/reth/src/commands/node/mod.rs:177]: no removing it

File: bin/reth/src/commands/node/mod.rs [hunk 2]
```diff
@@ -128,6 +132,77 @@ pub struct NodeCommand<Ext: RethCliExt = ()> {
...
+            txpool,
+            builder,
+            debug,
+            db,
+            dev,
+            pruning,
+            #[cfg(feature = "optimism")]
+            rollup,
+            ext,
+        } = self;
```
Comment (gakonst) [inline bin/reth/src/commands/node/mod.rs:163]: dont love this long destructure do we want a `From<NodeCommand> for NodeConfig` impl?

File: bin/reth/src/cli/mod.rs [hunk 3]
```diff
@@ -116,6 +147,20 @@ pub fn run() -> eyre::Result<()> {
     Cli::<()>::parse().run()
 }
 
+/// Convenience function for parsing CLI options, set up logging and run the chosen command.
+#[inline]
+pub fn run_ethereum() -> eyre::Result<()> {
```
Comment (gakonst) [inline bin/reth/src/cli/mod.rs:152]: where will this be used?

Comment (mattsse) [reply bin/reth/src/cli/mod.rs:152]: main.rs

File: bin/reth/src/commands/node/mod.rs [hunk 4]
```diff
@@ -124,56 +120,19 @@ pub struct NodeCommand<Ext: RethCliExt = ()> {
...
-    }
-
-    /// Execute `node` command
-    pub async fn execute(self, ctx: CliContext) -> eyre::Result<()> {
+impl<Ext: clap::Args + fmt::Debug> NodeCommand<Ext> {
+    /// Launches the node
+    ///
+    /// This transforms the node command into a node config and launches the node using the given
+    /// closure.
+    pub async fn execute2<L, Fut>(self, ctx: CliContext, launcher: L) -> eyre::Result<()>
```
Comment (onbjerg) [inline bin/reth/src/commands/node/mod.rs:131]:
```suggestion
    pub async fn execute<L, Fut>(self, ctx: CliContext, launcher: L) -> eyre::Result<()>
```

File: crates/consensus/auto-seal/tests/it/main.rs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-//! auto-mine consensus tests
-mod auto_mine;
+// //! auto-mine consensus tests
+// mod auto_mine;
```
Comment (onbjerg) [inline crates/consensus/auto-seal/tests/it/main.rs:2]: being added bck in this PR?

File: bin/reth/src/lib.rs [hunk 6]
```diff
@@ -26,7 +26,7 @@
 )]
 #![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
 
-pub mod builder;
+// pub mod builder;
```
Comment (onbjerg) [inline bin/reth/src/lib.rs:29]: being added back?

**Review by onbjerg — COMMENTED** [review]
looks good overall, but lot of modules commented out - are we adding these back?

File: examples/additional-rpc-namespace-in-cli/src/main.rs [hunk 7]
```diff
@@ -14,24 +14,39 @@
...
-struct MyRethCliExt;
-
-impl RethCliExt for MyRethCliExt {
-    /// This tells the reth CLI to install the `txpool` rpc namespace via `RethCliTxpoolExt`
-    type Node = RethCliTxpoolExt;
+    Cli::<RethCliTxpoolExt>::parse()
+        .run(|builder, args| async move {
+            let handle = builder
+                .with_types(EthereumNode::default())
+                .with_components(EthereumNode::components())
```
Comment (gakonst) [inline examples/additional-rpc-namespace-in-cli/src/main.rs:26]: Let's also add an alias for default setups `builder.network::<Ethereum>()`

File: examples/additional-rpc-namespace-in-cli/src/main.rs [hunk 8]
```diff
@@ -14,24 +14,39 @@
...
+                .extend_rpc_modules(move |ctx| {
```
Comment (gakonst) [inline examples/additional-rpc-namespace-in-cli/src/main.rs:27]: I assume people will want to extend these with functions vs doing it all in-line, so maybe we should refactor the closure logic to a helper function to make this more obvious i.e.

```
.extend_rpc_modules(move |ctx| extend_txpool(&mut ctx)?; println!("extensions configured"); Ok(()) })

File: crates/node-optimism/src/node.rs [hunk 9]
```diff
@@ -132,47 +154,75 @@ where
...
+        async move {
+            let Self { sequencer_http, disable_txpool_gossip } = self;
+            let mut network_config = ctx.network_config()?;
 
-        Ok(handle)
+            // When `sequencer_endpoint` is configured, the node will forward all transactions to a
+            // Sequencer node for execution and inclusion on L1, and disable its own txpool
+            // gossip to prevent other parties in the network from learning about them.
+            network_config.tx_gossip_disabled = disable_txpool_gossip;
+            network_config.optimism_network_config.sequencer_endpoint = sequencer_http;
```
Comment (gakonst) [inline crates/node-optimism/src/node.rs:219]: this is really nice

File: crates/node-optimism/src/node.rs [hunk 10]
```diff
@@ -132,47 +154,75 @@ where
...
-            ctx.chain_spec(),
-            payload_builder,
-        );
-        let (payload_service, payload_builder) =
-            PayloadBuilderService::new(payload_generator, ctx.provider().canonical_state_stream());
-
-        ctx.task_executor().spawn_critical("payload builder service", Box::pin(payload_service));
-
-        Ok(payload_builder)
+    ) -> impl Future<Output = eyre::Result<PayloadBuilderHandle<Node::Engine>>> + Send {
```
Comment (gakonst) [inline crates/node-optimism/src/node.rs:157]: Finding the req to return a future a bit odd but thinking about it

Comment (Rjected) [reply crates/node-optimism/src/node.rs:157]: this is desugared `async_trait` without the crate macro and without dynamic dispatch

**Review by Rjected — COMMENTED** [review]
I like this, lgtm pending examples migration and others review comments

File: crates/node-optimism/src/node.rs [hunk 11]
```diff
@@ -53,75 +55,95 @@ impl NodeTypes for OptimismNode {
...
-                reth_transaction_pool::maintain::maintain_transaction_pool_future(
-                    client,
-                    pool,
-                    chain_events,
+    fn build_pool(
+        self,
+        ctx: &BuilderContext<Node>,
+    ) -> impl Future<Output = eyre::Result<Self::Pool>> + Send {
+        async move {
+            let data_dir = ctx.data_dir();
```
Comment (gakonst) [inline crates/node-optimism/src/node.rs:71]: should we be reusing code here between node-optimism and node-ethereum?

File: crates/node-ethereum/src/node.rs [hunk 12]
```diff
@@ -132,31 +139,36 @@ where
         self,
         ctx: &BuilderContext<Node>,
         pool: Pool,
-    ) -> eyre::Result<PayloadBuilderHandle<Node::Engine>> {
-        let payload_builder = reth_ethereum_payload_builder::EthereumPayloadBuilder::default();
```
Comment (gakonst) [inline crates/node-ethereum/src/node.rs:136]: shoudl all these functions justbe made `async`?

File: crates/node-ethereum/src/node.rs [hunk 13]
```diff
@@ -63,58 +64,64 @@ where
...
-            ctx.task_executor().spawn_critical(
-                "txpool maintenance task",
-                reth_transaction_pool::maintain::maintain_transaction_pool_future(
-                    client,
-                    pool,
-                    chain_events,
+    fn build_pool(
+        self,
+        ctx: &BuilderContext<Node>,
+    ) -> impl Future<Output = eyre::Result<Self::Pool>> + Send {
```
Comment (gakonst) [inline crates/node-ethereum/src/node.rs:70]: see my point about reusing the code

File: crates/node-core/src/node_config.rs [hunk 14]
```diff
@@ -750,14 +738,6 @@ impl NodeConfig {
                 self.network.port + self.instance - 1,
             )));
 
-        // When `sequencer_endpoint` is configured, the node will forward all transactions to a
-        // Sequencer node for execution and inclusion on L1, and disable its own txpool
-        // gossip to prevent other parties in the network from learning about them.
-        #[cfg(feature = "optimism")]
-        let cfg_builder = cfg_builder
-            .sequencer_endpoint(self.rollup.sequencer_http.clone())
-            .disable_tx_gossip(self.rollup.disable_txpool_gossip);
```
Comment (gakonst) [inline crates/node-core/src/node_config.rs:759]: beautiful

File: crates/node-core/src/cli/mod.rs [hunk 15]
```diff
@@ -1,6 +1,6 @@
 //! Types for the CLI.
 
-pub mod components;
+// pub mod components;
 pub mod config;
-pub mod db_type;
-pub mod ext;
+// pub mod db_type;
+// pub mod ext;
```
Comment (gakonst) [inline crates/node-core/src/cli/mod.rs:6]: remove commented code

File: crates/node-builder/src/builder.rs [hunk 16]
```diff
@@ -718,7 +966,8 @@ impl<Node: FullNodeTypes> BuilderContext<Node> {
 
     /// Creates the [NetworkBuilder] for the node and blocks until it is ready.
     pub fn network_builder_blocking(&self) -> eyre::Result<NetworkBuilder<Node::Provider, (), ()>> {
-        self.executor.block_on(self.network_builder())
+        // TODO this shouldn't be required
```
Comment (gakonst) [inline crates/node-builder/src/builder.rs:969]: why not? what's the issue

File: crates/node-builder/src/builder.rs [hunk 17]
```diff
@@ -629,6 +655,217 @@ where
     }
 }
 
+/// A [NodeBuilder] with it's launch context already configured
```
Comment (gakonst) [inline crates/node-builder/src/builder.rs:658]: Pretty insane trait bounds here, recommend we document them well, but ultimately OK to have complex internals if they expose powerful API

File: crates/node-builder/src/builder.rs [hunk 18]
```diff
@@ -435,7 +449,19 @@ where
 
         // create pipeline
         let network_client = network.fetch_client().await?;
-        let (consensus_engine_tx, consensus_engine_rx) = unbounded_channel();
+        let (consensus_engine_tx, mut consensus_engine_rx) = unbounded_channel();
+
+        if let Some(store_path) = config.debug.engine_api_store.clone() {
```
Comment (gakonst) [inline crates/node-builder/src/builder.rs:454]: nice!

File: bin/reth/src/optimism.rs [hunk 19]
```diff
@@ -14,7 +21,37 @@ fn main() {
...
+                        safe_block_hash: head.hash(),
+                        finalized_block_hash: head.hash(),
+                    },
+                    None,
+                )
+                .await?;
+            }
+        }
+
+        node_exit_future.await
```
Comment (gakonst) [inline bin/reth/src/optimism.rs:53]: Mental note we should explore integrating Magi or parts of it https://github.com/a16z/magi/blob/82e2c51fb7a62c48d505d222b11d1ff5f784d129/src/engine/api.rs#L210

File: bin/reth/src/main.rs [hunk 20]
```diff
@@ -15,7 +19,14 @@ fn main() {
         std::env::set_var("RUST_BACKTRACE", "1");
     }
```
Comment (gakonst) [inline bin/reth/src/main.rs:20]: Can remove the cfg optimism feature flags above, I think

File: bin/reth/src/commands/node/mod.rs [hunk 21]
```diff
@@ -290,42 +244,49 @@ mod tests {
...
+        ])
+        .unwrap();
         assert_eq!(cmd.network.discovery.port, 300);
         assert_eq!(cmd.network.port, 99);
     }
 
     #[test]
     fn parse_metrics_port() {
-        let cmd = NodeCommand::<()>::try_parse_from(["reth", "--metrics", "9001"]).unwrap();
+        let cmd = NodeCommand::<NoArgs>::try_parse_from(["reth", "--metrics", "9001"]).unwrap();
```
Comment (gakonst) [inline bin/reth/src/commands/node/mod.rs:268]: The NoArgs explicit arg is not needed given above default config right

File: bin/reth/src/commands/node/mod.rs [hunk 22]
```diff
@@ -124,56 +120,19 @@ pub struct NodeCommand<Ext: RethCliExt = ()> {
...
+    /// Launches the node
+    ///
+    /// This transforms the node command into a node config and launches the node using the given
+    /// closure.
+    pub async fn execute2<L, Fut>(self, ctx: CliContext, launcher: L) -> eyre::Result<()>
+    where
+        L: FnOnce(WithLaunchContext<Arc<DatabaseEnv>, InitState>, Ext) -> Fut,
+        Fut: Future<Output = eyre::Result<()>>,
+    {
         let Self {
```
Comment (gakonst) [inline bin/reth/src/commands/node/mod.rs:136]: would still like to further simplify this fn setup

File: bin/reth/src/cli/mod.rs [hunk 23]
```diff
@@ -222,17 +206,18 @@ mod tests {
     fn override_trusted_setup_file() {
         // We already have a test that asserts that this has been initialized,
         // so we cheat a little bit and check that loading a random file errors.
-        let reth = Cli::<()>::try_parse_from(["reth", "node", "--trusted-setup-file", "README.md"])
-            .unwrap();
-        assert!(reth.run().is_err());
+        let reth =
+            Cli::<NoArgs>::try_parse_from(["reth", "node", "--trusted-setup-file", "README.md"])
+                .unwrap();
+        // assert!(reth.run().is_err());
```
Comment (gakonst) [inline bin/reth/src/cli/mod.rs:212]: commented out

**Review by gakonst — COMMENTED** [review]
big fan

File: crates/consensus/auto-seal/tests/it/main.rs (hunk 5 above)
Comment (mattsse) [reply crates/consensus/auto-seal/tests/it/main.rs:2]: need to update, but likely removing for this PR

File: bin/reth/src/lib.rs (hunk 6 above)
Comment (mattsse) [reply bin/reth/src/lib.rs:29]: this will be removed

File: examples/additional-rpc-namespace-in-cli/src/main.rs (hunk 7 above)
Comment (mattsse) [reply examples/additional-rpc-namespace-in-cli/src/main.rs:26]: great idea
I can make this work

File: crates/node-optimism/src/node.rs (hunk 11 above)
Comment (mattsse) [reply crates/node-optimism/src/node.rs:71]: eventually yes,
can introduce a node-common crate, but for now this is fine

I have another PR open for OP specific tx validator

File: examples/additional-rpc-namespace-in-cli/src/main.rs (hunk 8 above)
Comment (mattsse) [reply examples/additional-rpc-namespace-in-cli/src/main.rs:27]: I think this is fine, but can add more examples later

File: bin/reth/src/commands/node/mod.rs (hunk 21 above)
Comment (mattsse) [reply bin/reth/src/commands/node/mod.rs:261]: for some reason it still is

File: bin/reth/src/lib.rs (hunk 6 above)
Comment (Rjected) [inline bin/reth/src/lib.rs:29]: should be deleted now that the file is gone

File: examples/custom-node/src/main.rs [hunk 24]
```diff
@@ -190,10 +171,156 @@ impl EngineTypes for CustomEngineTypes {
...
+        let PayloadConfig {
+            initialized_block_env,
+            initialized_cfg,
+            parent_block,
+            extra_data,
+            attributes,
+            chain_spec,
+        } = config;
+        <reth_ethereum_payload_builder::EthereumPayloadBuilder  as PayloadBuilder<Pool,Client>>  ::build_empty_payload(client,
+                                                                                                                       PayloadConfig { initialized_block_env, initialized_cfg, parent_block, extra_data, attributes: attributes.0, chain_spec }
```
Comment (Rjected) [inline examples/custom-node/src/main.rs:314]: any way to make the formatting on this better?

**Review by Rjected — APPROVED** [review]
looks great, I just have nits. The UX in the examples are amazing

**Comment by Rjected** [comment]
and pending compile fixes

**Review by onbjerg — APPROVED** [review]
💥

File: crates/node-builder/src/builder.rs [hunk 25]
```diff
@@ -122,15 +134,40 @@ impl NodeBuilder<(), InitState> {
...
+    pub fn with_launch_context(
+        self,
+        task_executor: TaskExecutor,
+        data_dir: ChainPath<DataDirPath>,
+    ) -> WithLaunchContext<DB, InitState> {
+        WithLaunchContext { builder: self, task_executor, data_dir }
+    }
+
+    /// Creates an _ephemeral_ preconfigured node for testing purposes.
+    pub fn testing_node(
```
Comment (rkrasiuk) [inline crates/node-builder/src/builder.rs:154]: great

File: bin/reth/src/optimism.rs [hunk 26]
```diff
@@ -14,7 +21,34 @@ fn main() {
         std::env::set_var("RUST_BACKTRACE", "1");
     }
 
-    if let Err(err) = reth::cli::run() {
+    if let Err(err) = Cli::<RollupArgs>::parse().run(|builder, rollup_args| async move {
```
Comment (rkrasiuk) [inline bin/reth/src/optimism.rs:24]: magnificent

**Review by rkrasiuk — APPROVED** [review]
stellar work, ship it 🫡

**Pull request merged**